src/core/authorization/TokenLookup.ts

```diff
@@ -36,51 +60,87 @@ export class TokenLookup {
     return this.#getClientStorageEntry(resourceServer);
   }
 
-  get auth(): Token | null {
+  get auth(): StoredToken | null {
     return this.#getTokenForService(SERVICES.AUTH);
   }
 
-  get transfer(): Token | null {
+  get transfer(): StoredToken | null {
     return this.#getTokenForService(SERVICES.TRANSFER);
   }
 
-  get flows(): Token | null {
+  get flows(): StoredToken | null {
     return this.#getTokenForService(SERVICES.FLOWS);
   }
 
-  get groups(): Token | null {
+  get groups(): StoredToken | null {
     return this.#getTokenForService(SERVICES.GROUPS);
   }
 
-  get search(): Token | null {
+  get search(): StoredToken | null {
     return this.#getTokenForService(SERVICES.SEARCH);
   }
 
-  get timer(): Token | null {
+  get timer(): StoredToken | null {
     return this.#getTokenForService(SERVICES.TIMER);
   }
 
-  get compute(): Token | null {
+  get compute(): StoredToken | null {
     return this.#getTokenForService(SERVICES.COMPUTE);
   }
 
-  gcs(endpoint: string): Token | null {
+  gcs(endpoint: string): StoredToken | null {
     return this.getByResourceServer(endpoint);
   }
 
-  getByResourceServer(resourceServer: string): Token | null {
+  getByResourceServer(resourceServer: string): StoredToken | null {
     return this.#getClientStorageEntry(resourceServer);
   }
 
-  getAll(): Token[] {
+  getAll(): StoredToken[] {
     const entries = getStorage()
       .keys()
-      .reduce((acc: (Token | null)[], key) => {
+      .reduce((acc: (StoredToken | null)[], key) => {
         if (key.startsWith(this.#manager.storageKeyPrefix)) {
           acc.push(getTokenFromStorage(key));
         }
         return acc;
       }, []);
     return entries.filter(isToken);
   }
+
+  /**
+   * Add a token to the storage.
+   */
+  add(token: Token | TokenResponse) {
+    const created = Date.now();
+    const expires = token.expires_in ? created + token.expires_in * 1000 : null;
+    getStorage().set(`${this.#manager.storageKeyPrefix}${token.resource_server}`, {
+      ...token,
+      /**
+       * Add metadata to the token to track when it was created and when it expires.
+       */
+      __metadata: {
+        created,
+        expires,
+      },
+    });
+    if ('other_tokens' in token) {
+      token.other_tokens?.forEach((t) => {
+        this.add(t);
+      });
+    }
+  }
+
+  /**
+   * Determines whether or not a stored token is expired.
+   * @param token The token to check.
+   * @param augment An optional number of milliseconds to add to the current time when checking the expiration.
+   */
+  static isTokenExpired(token: StoredToken | null, augment: number = 0) {
+    /* eslint-disable no-underscore-dangle */
+    return token && token.__metadata?.expires
+      ? Date.now() + augment >= token.__metadata.expires
+      : true;
```

Comment:
Is the idea of having a fallback to `true` here to replace any old tokens that didn't get set with an `expires` value?

Reply:
Yeah, although thinking about it a bit more, I think I'm going to change the signature to `true | false | undefined` and  just let the caller decide what to do; In our applications, we'll likely do a refresh in that case.